source/equipment/equipment_api.cpp

```diff
@@ -0,0 +1,140 @@
+along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+Stay tuned using
+twitter @navitia
+IRC #navitia on freenode
+https://groups.google.com/d/forum/navitia
+www.navitia.io
+*/
+
+#include "equipment_api.h"
+#include "utils/paginate.h"
+
+#include "boost/range/algorithm.hpp"
+
+#include <tuple>
+#include <vector>
+#include <string>
+
+namespace navitia {
+namespace equipment {
+
+namespace {
+const std::string build_ptref_sa_filter(const std::string& line_uri, const std::string& filter) {
+    std::string sa_filter = "line.uri=" + line_uri;
+    if (filter.size()) {
+        sa_filter += " and (" + filter + ")";
+    }
+    return sa_filter;
+}
+
+const std::string build_ptref_sp_filter(const std::string& stop_area_uri, const std::string& filter) {
+    std::string sp_filter = "stop_area.uri=" + stop_area_uri;
+    if (filter.size()) {
+        sp_filter += " and (" + filter + ")";
+    }
+    return sp_filter;
+}
+
+void fill_equipment_to_pb(const equipment::EquipmentReportList& reports, PbCreator& pb_creator, int depth) {
+    for (const auto& report : reports) {
+        const type::Line* line = std::get<0>(report);
+        const auto& stop_areas_reports = std::get<1>(report);
+        auto pb_report = pb_creator.add_equipment_reports();
+        pb_creator.fill(line, pb_report->mutable_line(), depth);
+        for (const auto& sa_report : stop_areas_reports) {
+            const type::StopArea* sa = std::get<0>(sa_report);
+            const auto& stop_points = std::get<1>(sa_report);
+            auto sa_equip = pb_report->add_stop_area_equipments();
+            auto pb_sa = sa_equip->mutable_stop_area();
+            pb_creator.fill(sa, pb_sa, depth);
+            for (const type::StopPoint* sp : stop_points) {
+                auto pb_sp = pb_sa->add_stop_points();
+                pb_creator.fill(sp, pb_sp, depth);
+            }
+        }
+    }
+}
+}  // namespace
+
+EquipmentReports::EquipmentReports(const type::Data& data,
+                                   const std::string& filter,
+                                   int count,
+                                   int start_page,
+                                   const ForbiddenUris& forbidden_uris)
+    : data(data), filter(filter), count(count), start_page(start_page), forbidden_uris(forbidden_uris) {}
+
+EquipmentReportList EquipmentReports::get_paginated_equipment_report_list() {
+    const type::Indexes line_indices = ptref::make_query(type::Type_e::Line, filter, forbidden_uris, data);
+    const auto lines = data.get_data<type::Line>(line_indices);
+
+    total_lines = lines.size();
+    const auto paginated_lines = paginate(lines, count, start_page);
+
+    EquipmentReportList res;
+    for (const auto line : paginated_lines) {
+        const std::string sa_filter = build_ptref_sa_filter(line->uri, filter);
+        const type::Indexes sa_indexes = ptref::make_query(type::Type_e::StopArea, sa_filter, forbidden_uris, data);
```

Comment:
Another possible optim: request SP directly here, and store them directly, using `sp.stop_area`.
You could then remove the corresponding test.

Reply:
As suggested by @benoit-bst it's OK to move that optim' in the JIRA ticket. (not sure about the use of forbidden_uris though).
Please just keep track of it in the ticket if you delay it (and mention the ticket here in comment).

Reply:
@pbougue Doing so, you'll have a lot of repeated stop_area

Reply:
but, yes, we'll save a query...

Reply:
To precise a bit my thinking:
Do `const type::Indexes sp_indexes = ptref::make_query(type::Type_e::StopPoint, sa_filter, forbidden_uris, data);`
Then with this fill a `map<SA, StopAreaEquipment>` (or something like that, avoiding re-creation of SA)

This would save one query per line*SA-with-code (should not be negligible).

Reply:
After discussion, given the emergency and the fact that we already tracked a :mountain: of technical debt on our projects: No JIRA ticket on that, idea is tracked here in github in case someone searches through the code.
And we wait for perf issues to optimize it.